### PR TITLE
Graph coloring and MIS

### LINKED ISSUE #4230: Graph Coloring
Provide an algorithm for performing graph coloring.

Given a graph view, assign a unique integer to each vertex so that no two adjacent vertices have the same value.  Integers should be assigned from 0, we should use the minimum number of possible colors (integer values).

### LINKED ISSUE #4231: Maximal Independent Set
Provide an algorithm for computing Maximal Independent Set.

### PR DESCRIPTION
**Pull request by naimnv**
This PR implements 
- MNMG Graph Coloring
- MNMG MIS for general graphs

Closes #4230 
Closes #4231

### CODE REVIEW

File: cpp/include/cugraph/algorithms.hpp [hunk 1]
```diff
@@ -2349,6 +2349,25 @@ rmm::device_uvector<vertex_t> maximal_independent_set(
   graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
   raft::random::RngState& rng_state);
 
+/*
+ * @brief Find a Minimal Graph Coloring
+ *
```
Comment (ChuckHastings) [inline cpp/include/cugraph/algorithms.hpp:2354]: I feel like we need a bit more description of what this algorithm is doing.

File: cpp/include/cugraph/algorithms.hpp [hunk 2]
```diff
@@ -2349,6 +2349,25 @@ rmm::device_uvector<vertex_t> maximal_independent_set(
...
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Graph view object.
+ * @param rng_state The RngState instance holding pseudo-random number generator state.
+ * @return A device vector containing vertices found in the maximal independent set
```
Comment (ChuckHastings) [inline cpp/include/cugraph/algorithms.hpp:2362]: Not sure this description is correct.  Isn't this returning a color for each vertex?

File: cpp/src/community/detail/maximal_independent_moves.cuh [hunk 3]
```diff
@@ -0,0 +1,313 @@
+
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
```
Comment (ChuckHastings) [inline cpp/src/community/detail/maximal_independent_moves.cuh:3]: Is this a new file, or is this a copy/adaptation of an old file?

Comment (naimnv) [reply cpp/src/community/detail/maximal_independent_moves.cuh:3]: This is old file, just renamed.

File: cpp/include/cugraph/algorithms.hpp (hunk 1 above)
Comment (naimnv) [reply cpp/include/cugraph/algorithms.hpp:2354]: Added.

File: cpp/include/cugraph/algorithms.hpp (hunk 2 above)
Comment (naimnv) [reply cpp/include/cugraph/algorithms.hpp]: Changed.

File: cpp/include/cugraph/algorithms.hpp [hunk 4]
```diff
@@ -2349,6 +2349,25 @@ rmm::device_uvector<vertex_t> maximal_independent_set(
...
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Graph view object.
+ * @param rng_state The RngState instance holding pseudo-random number generator state.
+ * @return A device vector containing color for each vertex
+ */
+
```
Comment (seunghwak) [inline cpp/include/cugraph/algorithms.hpp:2364]: No need for an empty line here. This is inconsistent with the other algorithm declarations.

File: cpp/src/community/detail/maximal_independent_moves.cuh (hunk 3 above)
Comment (seunghwak) [reply cpp/src/community/detail/maximal_independent_moves.cuh:3]: Use `git mv` for just renaming a file.

Comment (seunghwak) [reply cpp/src/community/detail/maximal_independent_moves.cuh:3]: Skip reviewing this file assuming this is just renaming.

File: cpp/include/cugraph/algorithms.hpp (hunk 4 above)
Comment (seunghwak) [reply cpp/include/cugraph/algorithms.hpp:2364]: And based on https://en.wikipedia.org/wiki/Graph_coloring, `vertex_coloring` could be more descriptive name than just `coloring`.

Comment (seunghwak) [reply cpp/include/cugraph/algorithms.hpp:2364]: Is there any paper describing the algorithm behind your coloring implementation? If yes, better list it here.

File: cpp/src/components/coloring_impl.cuh [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+namespace detail {
+
+template <typename vertex_t, typename edge_t, bool multi_gpu>
+rmm::device_uvector<vertex_t> coloring(
+  raft::handle_t const& handle,
+  cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  raft::random::RngState& rng_state)
+{
+  auto const comm_rank = handle.get_comms().get_rank();
+  auto const comm_size = handle.get_comms().get_size();
```
Comment (seunghwak) [inline cpp/src/components/coloring_impl.cuh:38]: You'd better not assume that get_comms() will be valid in SG.

File: cpp/tests/CMakeLists.txt [hunk 6]
```diff
@@ -552,6 +548,14 @@ if(BUILD_CUGRAPH_MG_TESTS)
...
                     components/mg_weakly_connected_components_test.cpp)
 
+    ###############################################################################################
+    # - MG MIS tests ------------------------------------------------------------------------------
+    ConfigureTestMG(MG_MIS_TEST components/mg_mis_test.cu)
+
+    ###############################################################################################
+    # - MG Graph Coloring tests -------------------------------------------------------------------
+    ConfigureTestMG(MG_COLORING_TEST components/mg_coloring_test.cu)
+
```
Comment (seunghwak) [inline cpp/tests/CMakeLists.txt:558]: No SG tests?

Comment (naimnv) [reply cpp/tests/CMakeLists.txt:566]: Added.

File: cpp/src/components/coloring_impl.cuh (hunk 5 above)
Comment (naimnv) [reply cpp/src/components/coloring_impl.cuh]: Removed.

**Review by seunghwak — APPROVED** [review]
LGTM

**Review by ChuckHastings — APPROVED** [review]

**Pull request merged**